### ISSUE TITLE
Flaky test: DockerSwarmSuite.TestSwarmJoinPromoteLocked

### ISSUE THREAD

**Issue by tophj-ibm** [docker/docker#28987]
seen failing on arm and ppc64le.
https://jenkins.dockerproject.org/job/Docker%20Master%20(arm)/4014

```sh
07:35:18 FAIL: docker_cli_swarm_test.go:1091: DockerSwarmSuite.TestSwarmJoinPromoteLocked
07:35:18 
07:35:18 [def51f9347d96] waiting for daemon to start
07:35:18 [def51f9347d96] daemon started
07:35:18 [d5437d2e68f6d] waiting for daemon to start
07:35:18 [d5437d2e68f6d] daemon started
07:35:18 [d5437d2e68f6d] exiting daemon
07:35:18 [d5437d2e68f6d] waiting for daemon to start
07:35:18 [d5437d2e68f6d] daemon started
07:35:18 [dfad74b4f1ee0] waiting for daemon to start
07:35:18 [dfad74b4f1ee0] daemon started
07:35:18 [d5437d2e68f6d] exiting daemon
07:35:18 [d5437d2e68f6d] waiting for daemon to start
07:35:18 [d5437d2e68f6d] daemon started
07:35:18 [dfad74b4f1ee0] exiting daemon
07:35:18 [dfad74b4f1ee0] waiting for daemon to start
07:35:18 [dfad74b4f1ee0] daemon started
07:35:18 [dfad74b4f1ee0] exiting daemon
07:35:18 [dfad74b4f1ee0] waiting for daemon to start
07:35:18 [dfad74b4f1ee0] daemon started
07:35:18 [dfad74b4f1ee0] exiting daemon
07:35:18 [dfad74b4f1ee0] waiting for daemon to start
07:35:18 [dfad74b4f1ee0] daemon started
07:35:18 docker_cli_swarm_test.go:1171:
07:35:18     c.Assert(getNodeStatus(c, d3), checker.Equals, swarm.LocalNodeStateActive)
07:35:18 ... obtained swarm.LocalNodeState = "locked"
07:35:18 ... expected swarm.LocalNodeState = "active"
07:35:18 
07:35:18 [def51f9347d96] exiting daemon
07:35:18 [d5437d2e68f6d] exiting daemon
07:35:18 [dfad74b4f1ee0] exiting daemon
07:35:28 
07:35:28 ----------------------------------------------------------------------
```

and janky
https://jenkins.dockerproject.org/job/Docker-PRs/36140/consoleFull

```sh
03:04:24 ----------------------------------------------------------------------
03:04:24 FAIL: docker_cli_swarm_test.go:1091: DockerSwarmSuite.TestSwarmJoinPromoteLocked
03:04:24 
03:04:24 [d67d3fc49e76a] waiting for daemon to start
03:04:24 [d67d3fc49e76a] daemon started
03:04:24 [dfe2465b8d480] waiting for daemon to start
03:04:24 [dfe2465b8d480] daemon started
03:04:24 [dfe2465b8d480] exiting daemon
03:04:24 [dfe2465b8d480] waiting for daemon to start
03:04:24 [dfe2465b8d480] daemon started
03:04:24 [d333376d2e268] waiting for daemon to start
03:04:24 [d333376d2e268] daemon started
03:04:24 [dfe2465b8d480] exiting daemon
03:04:24 [dfe2465b8d480] waiting for daemon to start
03:04:24 [dfe2465b8d480] daemon started
03:04:24 [d333376d2e268] exiting daemon
03:04:24 [d333376d2e268] waiting for daemon to start
03:04:24 [d333376d2e268] daemon started
03:04:24 [d333376d2e268] exiting daemon
03:04:24 [d333376d2e268] waiting for daemon to start
03:04:24 [d333376d2e268] daemon started
03:04:24 docker_cli_swarm_test.go:1164:
03:04:24     c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
03:04:24 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc422568f20), Stderr:[]uint8(nil)} ("exit status 1")
03:04:24 ... out: Error response from daemon: swarm component could not be started: <nil>
03:04:24 
03:04:24 
03:04:24 [d67d3fc49e76a] exiting daemon
03:04:24 [dfe2465b8d480] exiting daemon
03:04:24 [d333376d2e268] exiting daemon
03:04:27 
03:04:27 ----------------------------------------------------------------------
```

**Comment by LK4D4** [comment]
Saw this on experimental
@cyli is it something fixed by your PR?

**Comment by cyli** [comment]
@LK4D4 I don't believe so - looking into these now

**Comment by LK4D4** [comment]
@cyli thanks!

**Comment by cyli** [comment]
These issues are caused because the test demotes a node, then immediately restarts it before it has a chance to get new TLS certs.  When it starts up again, the swarm node realizes it has been removed from the cluster, and stops.  Hence the second error message:

```
03:04:24 docker_cli_swarm_test.go:1164:
03:04:24     c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
03:04:24 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc422568f20), Stderr:[]uint8(nil)} ("exit status 1")
03:04:24 ... out: Error response from daemon: swarm component could not be started: <nil>
```

As a result, the node never becomes active - it's either locked or errored.

Ideally, if a node is demoted, it should restart itself as a worker, or even better, the swarmkit node should probably not shut down at all - it should just continue running as an agent only:  https://github.com/docker/swarmkit/issues/1782

**Comment by tophj-ibm** [comment]
@cyli I see. Thanks for looking into this!

**Comment by cyli** [comment]
Ok, so the second trace should be fixed by https://github.com/docker/swarmkit/pull/1790.

The first should be fixed by https://github.com/docker/docker/issues/29096

**Comment by ddingel** [comment]
Test [1422](https://jenkins.dockerproject.org/job/Docker%20Master%20(z)/1422/) failed today on z

**Comment by ddingel** [comment]
Run [1429](https://jenkins.dockerproject.org/job/Docker%20Master%20(z)/1429)  failed within `TestSwarmLockUnlockCluster` and the error pattern looks pretty much the same:

> 20:03:06 ----------------------------------------------------------------------
20:03:06 FAIL: docker_cli_swarm_test.go:922: DockerSwarmSuite.TestSwarmLockUnlockCluster
20:03:06 
20:03:06 [dda25dab351c2] waiting for daemon to start
20:03:06 [dda25dab351c2] daemon started
20:03:06 [d1f77520b29e8] waiting for daemon to start
20:03:06 [d1f77520b29e8] daemon started
20:03:06 [d2c814fa33746] waiting for daemon to start
20:03:06 [d2c814fa33746] daemon started
20:03:06 [d1f77520b29e8] exiting daemon
20:03:06 [d1f77520b29e8] waiting for daemon to start
20:03:06 [d1f77520b29e8] daemon started
20:03:06 [d1f77520b29e8] exiting daemon
20:03:06 [dda25dab351c2] exiting daemon
20:03:06 [dda25dab351c2] waiting for daemon to start
20:03:06 [dda25dab351c2] daemon started
20:03:06 [d2c814fa33746] exiting daemon
20:03:06 [d2c814fa33746] waiting for daemon to start
20:03:06 [d2c814fa33746] daemon started
20:03:06 [d1f77520b29e8] waiting for daemon to start
20:03:06 [d1f77520b29e8] daemon started
20:03:06 [d1f77520b29e8] exiting daemon
20:03:06 [d1f77520b29e8] waiting for daemon to start
20:03:06 [d1f77520b29e8] daemon started
20:03:06 [dda25dab351c2] exiting daemon
20:03:06 [dda25dab351c2] waiting for daemon to start
20:03:06 [dda25dab351c2] daemon started
20:03:06 [d2c814fa33746] exiting daemon
20:03:06 [d2c814fa33746] waiting for daemon to start
20:03:06 [d2c814fa33746] daemon started
20:03:06 docker_cli_swarm_test.go:987:
20:03:06     c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
20:03:06 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc4209405a0), Stderr:[]uint8(nil)} ("exit status 1")
20:03:06 ... out: Error response from daemon: swarm component could not be started: can't initialize raft node: decryption error using NACL secretbox
20:03:06 
20:03:06 
20:03:06 [dda25dab351c2] exiting daemon
20:03:06 [d1f77520b29e8] exiting daemon
20:03:06 [d2c814fa33746] exiting daemon

**Comment by cyli** [comment]
@ddingel Thank you!  I think https://github.com/docker/docker/pull/29096/files should fix [1422](https://jenkins.dockerproject.org/job/Docker%20Master%20(z)/1422/), and I'm looking into [1429](https://jenkins.dockerproject.org/job/Docker%20Master%20(z)/1429)

**Comment by ddingel** [comment]
New failing run [1448](https://jenkins.dockerproject.org/job/Docker%20Master%20(z)/1448/)

```
16:23:43 ----------------------------------------------------------------------
16:23:43 FAIL: docker_cli_service_update_test.go:14: DockerSwarmSuite.TestServiceUpdatePort
16:23:43 
16:23:43 [d5332be07db9e] waiting for daemon to start
16:23:43 [d5332be07db9e] daemon started
16:23:43 docker_cli_service_update_test.go:23:
16:23:43     waitAndAssert(c, defaultReconciliationTimeout, d.checkActiveContainerCount, checker.Equals, 1)
16:23:43 docker_utils.go:1576:
16:23:43     c.Assert(v, checker, args...)
16:23:43 ... obtained int = 0
16:23:43 ... expected int = 1
16:23:43 
16:23:43 [d5332be07db9e] exiting daemon
16:24:00 
16:24:00 -------------------------------------
```

Not sure if it also within the scope of this issue?

**Comment by tophj-ibm** [comment]
@ddingel check out #26601 for that issue

**Comment by cyli** [comment]
@ddingel I think [1429](https://jenkins.dockerproject.org/job/Docker%20Master%20(z)/1429) should be fixed by https://github.com/docker/swarmkit/pull/1798, which is slated for cherry-picking into docker.

**Comment by tophj-ibm** [comment]
closed via #29096
